### PR TITLE
Enumerable methods

### PR DESCRIPTION
**Pull request by damiecode**

### CODE REVIEW

**Comment by Torres-ssf** [comment]
### Pull Request Under Review By Sérgio (TSE)

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    arr = []
+    my_each do |i|
+      arr << if !arg.nil?
+               arg.call(i)
+             else
+               yield(i)
+             end
+    end
+    arr
+  end
```
Comment (Torres-ssf) [inline main.rb]: **[REQUIRE]**
My_map returns an Enumerator if no block is given

File: .rubocop.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  Max: 150
+Metrics/BlockLength:
+  ExcludedMethods: ["describe"]
+Metrics/ModuleLength:
+  Max: 150
+Metrics/CyclomaticComplexity:
+  Max: 100
+Metrics/PerceivedComplexity:
+  Max: 100
+
```
Comment (Torres-ssf) [inline .rubocop.yml]: **[REQUIRED]**
We need to remove these lines and use only the lines provided in the [original](https://github.com/microverseinc/linters-config/blob/master/ruby/.rubocop.yml) .rubocop.yml file provided by Microverse.

**Review by Torres-ssf — CHANGES_REQUESTED** [review]
### Great first submission :clap: 

But there are things that we need to change, we need to use the .rubocop.yml just like the original provided by Microverse. We also need to improve our `my_map`.

Please kindly submit for code review again after making these changes. Happy coding. :computer:

File: .rubocop.yml (hunk 2 above)
Comment (damiecode) [reply .rubocop.yml]: When I use the normal one, I keep getting errors that the complexity of some of the methods are too high. I don't know how else to write them

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    item_select = []
+    my_each { |i| item_select << i if yield(i) }
+    item_select
+  end
+
+  def my_all?
+    return my_each { |i| return false if yield(i) == false } unless block_given?
+
+    true
+  end
```
Comment (Ispirett) [inline main.rb]: My all should be able to take arguments please address this, also. You need to handles cases where a block of code is   not given.

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    false
+  end
+
+  def my_none?
+    return my_each { |i| return false if yield(i) } unless block_given?
+
+    true
+  end
+
```
Comment (Ispirett) [inline main.rb]: My all should be able to take arguments please address this, also. You need to handles cases where a block of code is not given.

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  def my_any?
+    return my_each { |i| return true if yield(i) } unless block_given?
+
+    false
+  end
+
```
Comment (Ispirett) [inline main.rb:63]: My any should be able to take arguments please address this, also. You need to handles cases where a block of code is not given.

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  def my_inject(value = self[0])
+    value ||= []
+    my_each do |element|
+      value = yield(value, element)
+    end
+    value
+  end
+end
```
Comment (Ispirett) [inline main.rb:128]: Your my_inject method needs some fixing :bug: 
enumerables my_inject combines all elements of enum by applying a binary operation, specified by a block
      Failure/Error: yield(self[i])
      
      NoMethodError:
        undefined method `[]' for 5..50:Range

**Review by Ispirett — CHANGES_REQUESTED** [review]
Great work with the changes you've made so far. :100:  :1st_place_medal: 
Onto the second half, there are some bugs :bug:  that need to die

Check out the comments below.
Remember to write descriptive commit messages.

File: main.rb (hunk 4 above)
Comment (damiecode) [reply main.rb]: Thanks for your review, please can you suggest what I should put if block is not given? All I can think of if puts an error. Thanks.
And what kind of arguments should they take

File: main.rb (hunk 6 above)
Comment (damiecode) [reply main.rb:128]: Hello, @Ispirett I have corrected the errors you highlighted and submitted for another review.

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    my_each { |i| item_select << i if yield(i) }
+    item_select
+  end
+
+  def my_all?(param = nil)
+    my_each { |i| return false if yield(i) == false } if block_given?
+    my_each { |_i| return false if item == false } if param.nil?
+
+    true
+  end
```
Comment (bolah2009) [inline main.rb]: What if `param` is not `nil`? :thinking: 

 - This method (`my_all?`) should only return true if the block never returns false or nil, when a block is given.
   ```ruby
   irb(main):006:0> [1,2,3].all?{|i| i > 0}
   => true
   irb(main):007:0> [1,2,3].my_all?{|i| i > 0}
   NameError (undefined local variable or method `item' for [1, 2, 3]:Array)
   ```

- When a pattern is passed this method should return true if every collection member matches this pattern. Note that pattern could be a `Class` (e.g. `Integer`, `String`) or `RegEx`
   ```ruby
   # When a Class pattern is passed
   irb(main):014:0> [1,2,'i'].all?(Integer)
   => false
   irb(main):015:0> [1,2,'i'].my_all?(Integer)
   => true

   # When a RegEx pattern is passed
   irb(main):018:0> %w[ant bat cat].all?(/d/)
   => false
   irb(main):019:0> %w[ant bat cat].my_all?(/d/)
   => true
   ```
   Hint: Use ` === ` to check if a member matches a patern.

### Find more details about [`all?` here](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F)

File: main.rb [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    return to_enum unless block_given?
+
+    item_select = []
+    my_each { |i| item_select << i if yield(i) }
+    item_select
+  end
+
+  def my_all?(param = nil)
+    my_each { |i| return false if yield(i) == false } if block_given?
+    my_each { |_i| return false if item == false } if param.nil?
```
Comment (bolah2009) [inline main.rb]: `item` is undefined here.

File: main.rb [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+  def my_any?(param = nil)
+    if block_given?
+      my_each { |i| return true if yield(i) }
+    elsif param.nil?
+      my_each { |i| return true if i }
+    else
+      my_each { |i| return true if i.class == param }
+    end
+    false
+  end
```
Comment (bolah2009) [inline main.rb]: - When a pattern is passed this method should return true if **any** collection member matches this pattern. Note that pattern could be a `Class` (e.g. `Integer`, `String`) or `RegEx`
   ```ruby
   # When a pattern is passed
   irb(main):023:0> [1,2,3,'s'].any?(2)
   => true
   irb(main):024:0> [1,2,3,'s'].my_any?(2)
   => false

   # When a RegEx pattern is passed
   irb(main):033:0> %w[ant dog cat].any?(/d/)
   => true
   irb(main):034:0> %w[ant dog cat].my_any?(/d/)
   => false
   ```
   Hint: Use ` === ` to check if a member matches a patern.

### Find more details about [`any?` here](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F)

File: main.rb [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+  def my_none?(param = nil)
+    if block_given?
+      my_each { |i| return false if yield(i) }
+    elsif param.nil?
+      my_each { |i| return false if i }
+    else
+      my_each { |i| return false if i.class == param }
+    end
+    true
+  end
```
Comment (bolah2009) [inline main.rb]: See comment on `my_any`.

### Find more details about [`none?` here](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F)

File: main.rb [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+             end
+    end
+    arr
+  end
+
+  def my_inject(*_args)
+    result = self[0]
+    self[1..-1].my_each { |i| result = yield(result, i) }
+    result
+  end
```
Comment (bolah2009) [inline main.rb]: #### :bug: 
- `my_inject` should be able to accept both an argument as initial value and a block:
   ```ruby
   irb(main):050:0> [1,2,3].inject(4) { |prod, n| prod * n }
   => 24
   irb(main):051:0> [1,2,3].my_inject(4) { |prod, n| prod * n }
   => 6
   ```
- `my_inject` should be able to accept only a `Symbol` as an operator (e.g `:+`, `:*`):
   ```ruby
   irb(main):055:0> [1,2,3].inject(:+)
   => 6
   irb(main):056:0> [1,2,3].my_inject(:+)
   LocalJumpError (no block given (yield))
   ```

- `my_inject` should be able to accept two arguments, first as an initial value and the other as a `Symbol`:

   ```ruby
   irb(main):057:0> [1,2,3].inject(5,:+)
   => 11
   irb(main):058:0> [1,2,3].my_inject(5,:+)
   LocalJumpError (no block given (yield))
   ```

### Find more details about [`inject?` here](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject)

**Review by bolah2009 — CHANGES_REQUESTED** [review]
## Hey, @damiecode 

Please check the comments below. Reach out on slack if you have issues implementing the suggested changes.

#### :recycle:  Make these changes and submit another code review request. :+1: 

Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)

###### PS: Note that comments tagged as `OPTIONAL` are not part of the minimum requirements to approve this PR

File: .rubocop.yml [hunk 12]
```diff
@@ -0,0 +1,36 @@
+inherit_from: .rubocop_todo.yml
```
Comment (bolah2009) [inline .rubocop.yml]: Can you delete this?
```suggestion
```

File: main.rb [hunk 13]
```diff
@@ -0,0 +1,139 @@
+# frozen_string_literal: true
+
+module Enumerable # rubocop:disable Metrics/ModuleLength
```
Comment (bolah2009) [inline main.rb]: Let's move `# rubocop:disable Metrics/ModuleLength` to the top of the module.
```suggestion
# rubocop:disable Metrics/ModuleLength
module Enumerable
```

File: main.rb [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+  arr.my_inject { |a, b| a * b }
+end
+
+var1 = multiply_els([2, 4, 5])
+print var1
+puts ''
+
+block = proc { |string| string.upcase }
+var2 = %w[a b c].my_map(block)
+print var2
```
Comment (bolah2009) [inline main.rb:141]: We can move this to the readme instead. :thinking: 

[Optional]

File: main.rb [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+
+  def my_select
+    return to_enum unless block_given?
+
+    item_select = []
+    my_each { |i| item_select << i if yield(i) }
+    item_select
+  end
+
+  def my_all?(param = nil)
```
Comment (bolah2009) [inline main.rb:34]: To fix linter issues, you can use inline disable rules.

```suggestion
 # rubocop:disable  Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
  def my_all?(param = nil)
```

File: main.rb [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+    arr[0..-1].my_each do |i|
+      operand = if symbol
+                  operand.send(symbol, i)
+                else
+                  yield(operand, i)
+                end
+    end
+    operand
+  end
+end
```
Comment (bolah2009) [inline main.rb:128]: Let's enable it back here.
```suggestion
end 
# rubocop:enable  Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
```

**Review by bolah2009 — APPROVED** [review]
### Awesome work @damiecode :+1: 

This project is approved. 

Please fix the linter issues before merging.

File: main.rb (hunk 16 above)
Comment (damiecode) [reply main.rb:128]: Thank you very much @bolah2009

**Pull request merged**